nixos/modules/config/timezone.nix

```diff
@@ -40,13 +43,14 @@ in
     # This way services are restarted when tzdata changes.
     systemd.globalEnvironment.TZDIR = tzdir;
 
-    environment.etc.localtime =
-      { source = "/etc/zoneinfo/${config.time.timeZone}";
-        mode = "direct-symlink";
-      };
-
-    environment.etc.zoneinfo.source = tzdir;
+    systemd.services.systemd-timedated.environment = lib.optionalAttrs (config.time.timeZone != null) { NIXOS_STATIC_TIMEZONE = "1"; };
 
+    environment.etc = {
+      zoneinfo.source = tzdir;
+    } // (if config.time.timeZone == null then {} else
```

Comment:
It would be slightly simpler to use [optionalAttrs](https://github.com/NixOS/nixpkgs/blob/5989515b9410f3d557e1aba037325a218f4338f5/lib/attrsets.nix#L321) here.